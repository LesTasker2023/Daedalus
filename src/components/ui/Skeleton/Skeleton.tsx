import { cn } from '@/lib/utils';
import * as React from 'react';

type SkeletonProps = React.HTMLAttributes<HTMLDivElement>;

const Skeleton = React.forwardRef<HTMLDivElement, SkeletonProps>(({ className, ...props }, ref) => {
  return (
    <div ref={ref} className={cn('animate-pulse rounded-md bg-primary/10', className)} {...props} />
  );
});
Skeleton.displayName = 'Skeleton';

export { Skeleton };
export type { SkeletonProps };
