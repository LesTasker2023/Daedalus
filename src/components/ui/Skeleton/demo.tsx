'use client';

import React, { useState } from 'react';
import { Skeleton } from '.';

export default function SkeletonDemo() {
  const [state, setState] = useState(false);

  return (
    <div className="space-y-6 p-6 max-w-2xl mx-auto">
      <div className="space-y-2">
        <h1 className="text-2xl font-bold">Skeleton Component</h1>
        <p className="text-muted-foreground">Loading state placeholders</p>
      </div>

      <div className="space-y-4">
        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Basic Usage</h3>
          <div className="border rounded-lg p-4 bg-background">
            <Skeleton>Skeleton Example</Skeleton>
          </div>
        </div>

        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Interactive</h3>
          <div className="border rounded-lg p-4 bg-background space-y-2">
            <Skeleton onClick={() => setState(!state)}>
              Click me {state ? '(Active)' : '(Inactive)'}
            </Skeleton>
            <p className="text-sm text-muted-foreground">State: {state ? 'Active' : 'Inactive'}</p>
          </div>
        </div>
      </div>
    </div>
  );
}
