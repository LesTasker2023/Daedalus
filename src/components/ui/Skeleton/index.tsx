export { default as SkeletonDemo } from './demo';
export * from './Skeleton';
export type { SkeletonProps } from './Skeleton';
