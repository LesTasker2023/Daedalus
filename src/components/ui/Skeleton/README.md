# Skeleton Component

Enterprise-grade Skeleton component with comprehensive testing and accessibility.

## Usage

```tsx
import { Skeleton } from '@/components/ui/Skeleton';

<Skeleton />;
```

## Props

See TypeScript definitions for complete props interface.

## Accessibility

- WCAG 2.1 AA compliant
- Full keyboard navigation support
- Screen reader optimized
