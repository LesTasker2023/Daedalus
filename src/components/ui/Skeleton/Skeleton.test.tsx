import { render, screen } from '@testing-library/react';
import * as React from 'react';
import { beforeEach, describe, expect, it } from 'vitest';
import { Skeleton } from './Skeleton';

describe('Skeleton Component', () => {
  beforeEach(() => {
    // Reset any state before each test
  });

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(<Skeleton>Default</Skeleton>);
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches all variants snapshot', () => {
      const { container } = render(
        <div data-testid="variants-container">
          <Skeleton variant="default">Default</Skeleton>
          <Skeleton variant="destructive">Destructive</Skeleton>
          <Skeleton variant="outline">Outline</Skeleton>
          <Skeleton variant="secondary">Secondary</Skeleton>
        </div>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches all sizes snapshot', () => {
      const { container } = render(
        <div data-testid="sizes-container">
          <Skeleton size="sm">Small</Skeleton>
          <Skeleton size="default">Default</Skeleton>
          <Skeleton size="lg">Large</Skeleton>
        </div>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches disabled state snapshot', () => {
      const { container } = render(<Skeleton disabled>Disabled</Skeleton>);
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches loading state snapshot', () => {
      const { container } = render(<Skeleton loading>Loading</Skeleton>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Rendering', () => {
    it('renders without crashing', () => {
      render(<Skeleton data-testid="skeleton" />);
      expect(screen.getByTestId('skeleton')).toBeInTheDocument();
    });

    it('applies custom className', () => {
      render(<Skeleton className="custom-class" data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toHaveClass('custom-class');
    });

    it('has default skeleton styling', () => {
      render(<Skeleton data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toHaveClass('animate-pulse', 'rounded-md', 'bg-primary/10');
    });
  });

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      render(<Skeleton aria-label="Loading content" data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toHaveAttribute('aria-label', 'Loading content');
    });

    it('can be used with loading state semantics', () => {
      render(<Skeleton aria-live="polite" aria-busy="true" data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toHaveAttribute('aria-live', 'polite');
      expect(skeleton).toHaveAttribute('aria-busy', 'true');
    });
  });

  describe('Ref Forwarding', () => {
    it('forwards ref correctly', () => {
      const ref = React.createRef<HTMLDivElement>();
      render(<Skeleton ref={ref} data-testid="skeleton" />);
      expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });
  });

  describe('Edge Cases', () => {
    it('handles null props gracefully', () => {
      render(<Skeleton data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toBeInTheDocument();
    });

    it('supports various sizes through className', () => {
      render(<Skeleton className="h-4 w-full" data-testid="skeleton" />);
      const skeleton = screen.getByTestId('skeleton');
      expect(skeleton).toHaveClass('h-4', 'w-full');
    });
  });
});
