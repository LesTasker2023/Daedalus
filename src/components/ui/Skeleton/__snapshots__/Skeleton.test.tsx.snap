// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton > Snapshots > matches default snapshot 1`] = `
<div
  class="animate-pulse rounded-md bg-primary/10"
  data-testid="skeleton"
>
  Test content
</div>
`;
